clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.h

```diff
@@ -0,0 +1,62 @@
+//===- SYCLOffloadWrapper.h --r----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_TOOLS_CLANG_LINKER_WRAPPER_SYCL_OFFLOAD_WRAPPER_H
+#define LLVM_CLANG_TOOLS_CLANG_LINKER_WRAPPER_SYCL_OFFLOAD_WRAPPER_H
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/PropertySetIO.h"
+
+#include <string>
+#include <utility>
+
+// SYCL binary image formats supported.
+enum class SYCLBinaryImageFormat {
+  BIF_None,   // Undetermined Image kind
+  BIF_Native, // Native Image kind
+  BIF_SPIRV,  // SPIR-V
+  BIF_LLVMBC  // LLVM bitcode
+};
+
+SYCLBinaryImageFormat getBinaryImageFormatFromString(llvm::StringRef S);
+
+struct SYCLImage {
+  std::vector<char> Image;
+  std::optional<llvm::util::PropertySetRegistry> PropertyRegistry;
+
+  // Note: Must be null-terminated.
+  std::optional<std::vector<char>> Entries;
```

Comment:
Why are those optional? I believe that both containers are capable of expressing an empty state